pandas/core/internals/managers.py

```diff
@@ -1936,6 +1936,19 @@ def _compare_or_regex_search(a, b, regex=False):
     is_a_array = isinstance(a, np.ndarray)
     is_b_array = isinstance(b, np.ndarray)
 
+    def _get_nan_value(x):
```

Comment:
this is not a good strategy to do at all

rather NA should simply be handled

Reply:
Indeed, the behaviour of pd.NA seems problematic. For example:
```
>>> 1 == pd.NA
<NA>
>>> 1 == np.nan
False
```
Both of the cases should return a boolean value, right?

Reply:
@jreback: Do you mean something like the following?
```python
    from pandas import NA
    is_pd_NA = np.vectorize(lambda x: x is NA)
    if is_a_array:
        a = np.where(is_pd_NA(a), _get_nan_value(a), a)
    if is_b_array:
        b = np.where(is_pd_NA(b), _get_nan_value(b), b)
```

Reply:
so this is what you need to do.

compute the result (line 1956), then apply a mask

your mask is like this (may need adjustments if we have scalars)
```
mask = isna(a) | isna(b)

# import this at the top
na_value = pandas.core.dtypes.na_value_for_dtype(compat=False)

result[mask] = na_value
```

Reply:
Shouldn't the mask also be applied to both arrays before the computation of the result takes place? Otherwise the element-wise comparison is going to fail and the value of the result will be a scalar instead of an array. I will commit the above changes so that you can have a look.